### PR TITLE
Update/calendar new features

### PR DESCRIPTION
**Pull request by paulmatchett**

### CODE REVIEW

File: web-app/src/components/common/Input/index.js [hunk 1]
```diff
@@ -126,6 +126,11 @@ const Input = props => {
 
     if (htmlAttrs.disabled) {
       styles.push('isDisabled');
+    } else {
+      var index = styles.indexOf('isDisabled');
```
Comment (William-Unosquare) [inline web-app/src/components/common/Input/index.js]: ~~var~~ let ;)

File: web-app/src/pages/Dashboard/container.js [hunk 2]
```diff
@@ -33,46 +36,76 @@ const DashboardContainer = Wrapped =>
...
+              forename: userDetails.forename,
+              surname: userDetails.surname,
+              email: userDetails.email,
+            },
+            {
+              forename: 'Jim',
+              surname: 'Beam',
+              email: 'jim@gmail.com',
+            },
+          ];
```
Comment (William-Unosquare) [inline web-app/src/pages/Dashboard/container.js]: Remove all this placeholder stuff if you suspect it won't be needed when this gets merged in with the new backend

File: web-app/src/pages/Dashboard/container.js [hunk 3]
```diff
@@ -33,46 +36,76 @@ const DashboardContainer = Wrapped =>
...
+          for (let hol of usersHolidays) {
+            var eventId = this.getRandomItem(eventArray);
+            var user = this.getRandomItem(personArray);
```
Comment (William-Unosquare) [inline web-app/src/pages/Dashboard/container.js]: ~~var~~ let

**Review by William-Unosquare — CHANGES_REQUESTED** [review]
This is a lot to go through so there could be more issues. I'll look through this again sooner to when it will be merged. For now, there are a few small things you could change.

File: web-app/src/components/common/Input/index.js (hunk 1 above)
Comment (paulmatchett) [reply web-app/src/components/common/Input/index.js]: yep, your right. changed

File: web-app/src/pages/Dashboard/container.js (hunk 2 above)
Comment (paulmatchett) [reply web-app/src/pages/Dashboard/container.js]: all dummy data removed. will default to test data in axios middleware

File: web-app/src/pages/Dashboard/container.js (hunk 3 above)
Comment (paulmatchett) [reply web-app/src/pages/Dashboard/container.js]: yep, again... your right. changed

File: web-app/src/components/BookingCalendar/container.js [hunk 4]
```diff
@@ -15,15 +16,23 @@ const BookingCalendarContainer = Wrapped =>
...
+          isEventBeingUpdated: false,
+          isHalfday: false,
+          isWFH: false,
+          start: moment(start),
+        };
+        this.props.updateBookingAndDuration(booking);
+      } else {
+        Swal({
+          text: 'Unable to select past dates',
+          title: 'Sorry',
```
Comment (William-Unosquare) [inline web-app/src/components/BookingCalendar/container.js]: Good use-case for the toast when it's merged

File: web-app/src/components/BookingCalendar/container.js [hunk 5]
```diff
@@ -33,6 +42,11 @@ const BookingCalendarContainer = Wrapped =>
           isEventBeingUpdated: true,
         };
         this.props.updateBookingAndDuration(updatedBooking);
+      } else {
+        Swal({
+          text: 'Unable to update other peoples events',
+          title: 'Sorry',
```
Comment (William-Unosquare) [inline web-app/src/components/BookingCalendar/container.js]: Another good use-case for toast 🍞

File: web-app/src/components/BookingModalForm/index.js [hunk 6]
```diff
@@ -40,70 +43,104 @@ const BookingModalForm = props => {
     }
   };
 
-  const updateVisability = () => {
-    return Math.floor(
-      moment.duration(formData.end.diff(formData.start)).asDays(),
-    ) === 0
-      ? false
-      : true;
+  const calculateDaysNotice = () => {
```
Comment (William-Unosquare) [inline web-app/src/components/BookingModalForm/index.js]: I can see this method being used in other places in the app - maybe abstract it out into a util?

File: web-app/src/utilities/dates.js [hunk 7]
```diff
@@ -30,3 +30,21 @@ export const getTotalDaysInEventArrayWithStatus = (events, status) => {
...
+export const getDuration = (startDate, endDate) => {
+  return Math.floor(moment.duration(endDate.diff(startDate)).asDays() + 1);
+};
+
+export const getDurationNotice = (startDate, endDate) => {
+  let days = getDuration(startDate, endDate);
+  let businessDays = 0;
+  let date = new moment(startDate);
+  while (days > 0) {
+    if (date.isoWeekday() !== 6 && date.isoWeekday() !== 7) {
```
Comment (William-Unosquare) [inline web-app/src/utilities/dates.js]: Any reason not to use `if (date.isoWeekday() < 6)` here?

**Review by William-Unosquare — DISMISSED** [review]
A couple of small changes here, but fantastic work dude (ignore the toast stuff, it will be good to come back to) 👍

File: web-app/src/components/BookingModalForm/index.js (hunk 6 above)
Comment (paulmatchett) [reply web-app/src/components/BookingModalForm/index.js]: method has been abstracted out to utilities/dates.js

File: web-app/src/components/BookingModal/BookingModalForm/index.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+      const daysNotice = calculateDaysNotice(daysRequested);
+      let error;
+      if (daysRequested == 0) {
+        error = {
+          message: 'Are you sure you wish to request a day off over a weekend?',
+        };
+      } else {
+        error = {
+          message: `You need to give ${daysNotice} working/business days notice to request ${daysRequested} ${
```
Comment (William-Unosquare) [inline web-app/src/components/BookingModal/BookingModalForm/index.js]: Maybe change the wording of this message slightly. From "You need" to "You should" so it doesn't sound like such a hard requirement.

File: web-app/src/components/BookingCalendar/container.js [hunk 9]
```diff
@@ -28,9 +29,9 @@ const BookingCalendarContainer = Wrapped =>
         };
         this.props.updateBookingAndDuration(booking);
       } else {
-        Swal({
-          text: 'Unable to select past dates',
-          title: 'Sorry',
+        Toast({
+          type: 'Sorry',
```
Comment (William-Unosquare) [inline web-app/src/components/BookingCalendar/container.js]: The 'type' option is for the icon. Use 'info'

File: web-app/src/components/BookingCalendar/container.js [hunk 10]
```diff
@@ -43,8 +44,8 @@ const BookingCalendarContainer = Wrapped =>
         };
         this.props.updateBookingAndDuration(updatedBooking);
       } else {
-        Swal({
-          text: 'Unable to update other peoples events',
+        Toast({
+          type: 'Unable to update other peoples events',
```
Comment (William-Unosquare) [inline web-app/src/components/BookingCalendar/container.js]: Again, type option is for the icon. Use `info`

File: web-app/src/components/BookingCalendar/container.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+    onSelectSlot = ({ start, end }) => {
+      const today = new moment();
+      if (moment(start).isAfter(today.add(-1, 'days'))) {
+        let booking = {
+          title: null,
+          end: moment(end),
+          eventType: {
+            eventTypeId: 1,
+            description: 'annual leave',
```
Comment (William-Unosquare) [inline web-app/src/components/BookingCalendar/container.js]: Capitalize the description to keep it consistent with the others

File: web-app/src/components/BookingModal/BookingModalForm/container.js [hunk 12]
```diff
@@ -1,13 +1,16 @@
...
+import {
+  updateHoliday,
+  requestHoliday,
+} from '../../../services/holidayService';
 
 export default Wrapped =>
   class extends React.Component {
     static propTypes = {
-      employeeId: PT.string.isRequired,
+      employeeId: PT.any.isRequired,
```
Comment (William-Unosquare) [inline web-app/src/components/BookingModal/BookingModalForm/container.js]: Why is this set to any? Surely that's not right.

File: web-app/src/components/BookingModal/index.js [hunk 13]
```diff
@@ -63,9 +57,9 @@ BookingModal.propTypes = {
   closeModal: PT.func.isRequired,
   showModal: PT.bool.isRequired,
   booking: PT.object.isRequired,
-  employeeId: PT.string.isRequired,
+  employeeId: PT.any.isRequired,
```
Comment (William-Unosquare) [inline web-app/src/components/BookingModal/index.js]: `any` Here too :P

File: web-app/src/components/Legend/index.jsx [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+      <Column>
+        <Form formData={selectedEmployee} formStatus={formStatus}>
+          <Input
+            type="select"
+            htmlAttrs={{
+              name: 'employee',
+              options: employees,
+            }}
+            value={selectedEmployee.employeeId}
+            label="FILITER BY EMPLOYEE:"
```
Comment (William-Unosquare) [inline web-app/src/components/Legend/index.jsx]: Spelling

File: web-app/src/components/common/Event/index.js [hunk 15]
```diff
@@ -3,12 +3,30 @@ import { PropTypes as PT } from 'prop-types';
...
+    id = eventStatusId;
+  } else if (eventTypeId === 2) {
+    // working from home
+    id = 4;
+  } else if (eventTypeId === 3) {
+    // sick leave
+    id = 5;
+  } else if (eventTypeId === 4) {
+    // work related travel
+    id = 4;
```
Comment (William-Unosquare) [inline web-app/src/components/common/Event/index.js]: We should have these as constants somewhere. I know there's comments but it's still awkward to read this code.

File: web-app/src/pages/Dashboard/container.js [hunk 16]
```diff
@@ -33,109 +29,165 @@ const DashboardContainer = Wrapped =>
       const userId = localStorage.getItem('user_id');
       getUserProfile(userId)
         .then(response => {
-          this.setState({
```
Comment (William-Unosquare) [inline web-app/src/pages/Dashboard/container.js:36]: Do we need to do this? The user profile is stored in Redux

File: web-app/src/pages/Dashboard/container.js [hunk 17]
```diff
@@ -33,109 +29,165 @@ const DashboardContainer = Wrapped =>
...
 
-    getTakenHolidays = userId => {
-      getHolidays(userId)
+    createBookingObj = event => {
+      const startDate = moment(event.startDate);
+      const endDate = moment(event.endDate);
+
+      const eventType = {
+        eventTypeId: 1,
+        description: 'annual leave',
```
Comment (William-Unosquare) [inline web-app/src/pages/Dashboard/container.js]: Capitalize similar to the other description below

File: web-app/src/pages/Dashboard/container.js [hunk 18]
```diff
@@ -33,109 +29,165 @@ const DashboardContainer = Wrapped =>
...
 
-    onSelectEvent = booking => {
-      const updatedBooking = {
-        ...booking,
-        isEventBeingUpdated: true,
-      };
-      this.getDuration(updatedBooking);
+    onFilterEvents = eventStatusId => {
+      let updatedFilterEvents = [...this.state.filterEvents];
+      if (eventStatusId !== undefined) {
```
Comment (William-Unosquare) [inline web-app/src/pages/Dashboard/container.js:136]: Picky, but `if (eventStatusId) {` should be enough here, I think

File: web-app/src/pages/Dashboard/index.jsx [hunk 19]
```diff
@@ -1,79 +1,62 @@
...
   booking: PT.object.isRequired,
   closeModal: PT.func.isRequired,
+  onUpdateEvents: PT.func.isRequired,
+  onUpdateEmployee: PT.func.isRequired,
   showModal: PT.bool.isRequired,
   takenHolidays: PT.array.isRequired,
-  getDuration: PT.func.isRequired,
+  updateBookingAndDuration: PT.func.isRequired,
+  updateTakenHolidays: PT.func.isRequired,
+  employeeId: PT.any.isRequired,
```
Comment (William-Unosquare) [inline web-app/src/pages/Dashboard/index.jsx]: `any` again

File: web-app/src/utilities/AxiosInstance.js [hunk 20]
```diff
@@ -18,36 +18,6 @@ instance.interceptors.response.use(function(response) {
...
-        startDate: [2014, 1, 1],
-        countryId: 1,
-        employeeRoleId: 2,
-        employeeStatusId: 2,
-      };
-
-      holidays[index].start = new moment(
-        holidays[index].startDate,
-        'YYYY-MM-DD',
-      );
```
Comment (William-Unosquare) [inline web-app/src/utilities/AxiosInstance.js]: Is there a reason you've removed the creation of moments here?

**Review by William-Unosquare — CHANGES_REQUESTED** [review]
Quite a few things to change here as to be expected with such a big PR. But overall great work 👍

File: web-app/src/components/BookingModal/index.js (hunk 13 above)
Comment (paulmatchett) [reply web-app/src/components/BookingModal/index.js]: changes to number any it worked this time!

File: web-app/src/components/Legend/index.jsx (hunk 14 above)
Comment (paulmatchett) [reply web-app/src/components/Legend/index.jsx]: updated

File: web-app/src/components/common/Event/index.js (hunk 15 above)
Comment (paulmatchett) [reply web-app/src/components/common/Event/index.js]: using holiday status to assign to each id now

File: web-app/src/pages/Dashboard/container.js (hunk 18 above)
Comment (paulmatchett) [reply web-app/src/pages/Dashboard/container.js:136]: yeah, your right

File: web-app/src/pages/Dashboard/container.js (hunk 16 above)
Comment (paulmatchett) [reply web-app/src/pages/Dashboard/container.js:36]: For some reason the user doesn't exist in the redux store. When the page loads. not sure what the issue is

File: web-app/src/utilities/AxiosInstance.js (hunk 20 above)
Comment (paulmatchett) [reply web-app/src/utilities/AxiosInstance.js]: on getting holiday events in dashboard container, i'm removing properties not required, so it was just as handy to sort the moment dates out there too.

Comment (William-Unosquare) [reply web-app/src/utilities/AxiosInstance.js]: @paulmatchett But by removing it here you're removing it from anywhere else that isn't the dashboard that makes the holiday calls.

**Review by William-Unosquare — APPROVED** [review]

**Comment by William-Unosquare** [comment]
Happpy to get this merged now because it's massive. There's a couple of things to be mindful to look into. Namely that unrequired user profile request instead of using redux.

**Pull request merged**